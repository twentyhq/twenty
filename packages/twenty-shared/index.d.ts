// Ambient module declarations to help TypeScript resolve subpath types in workspace consumers

declare module 'twenty-shared' {
  export * from './dist/index';
}

declare module 'twenty-shared/constants' {
  export * from './dist/constants/index';
}

declare module 'twenty-shared/testing' {
  export * from './dist/testing/index';
}

declare module 'twenty-shared/translations' {
  export * from './dist/translations/index';
}

declare module 'twenty-shared/types' {
  export * from './dist/types/index';
}

declare module 'twenty-shared/utils' {
  export * from './dist/utils/index';
}

declare module 'twenty-shared/workflow' {
  export * from './dist/workflow/index';
}

declare module 'twenty-shared/workspace' {
  export * from './dist/workspace/index';
}


