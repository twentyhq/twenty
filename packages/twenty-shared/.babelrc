{
  "$schema": "https://json.schemastore.org/babelrc.json",
  "presets": ["@babel/preset-env", "@babel/preset-typescript"],
  "plugins": [
    [
      "module-resolver",
      {
        "root": ["./"],
        "alias": {
          "@": "./src"
        }
      }
    ]
  ]
}
