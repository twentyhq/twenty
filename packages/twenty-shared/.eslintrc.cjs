module.exports = {
  extends: ['../../.eslintrc.global.cjs'],
  overrides: [
    {
      files: ['**/*.ts'],
      parserOptions: {
        project: ['packages/twenty-shared/tsconfig.*.json'],
      },
      rules: {
        '@nx/dependency-checks': 'error',
      },
    },
  ],
};
