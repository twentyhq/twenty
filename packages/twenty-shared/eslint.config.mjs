import js from '@eslint/js';
import nxPlugin from '@nx/eslint-plugin';
import typescriptEslint from '@typescript-eslint/eslint-plugin';
import typescriptParser from '@typescript-eslint/parser';
import importPlugin from 'eslint-plugin-import';
import linguiPlugin from 'eslint-plugin-lingui';
import preferArrowPlugin from 'eslint-plugin-prefer-arrow';
import prettierPlugin from 'eslint-plugin-prettier';
import unicornPlugin from 'eslint-plugin-unicorn';
import unusedImportsPlugin from 'eslint-plugin-unused-imports';
import jsoncParser from 'jsonc-eslint-parser';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export default [
  // Base JavaScript configuration
  js.configs.recommended,

  // Lingui recommended rules
  linguiPlugin.configs['flat/recommended'],

  // Global ignores
  {
    ignores: [
      '**/node_modules/**',
    ],
  },

  // Base configuration for all files
  {
    files: ['**/*.{js,jsx,ts,tsx}'],
    plugins: {
      'prettier': prettierPlugin,
      'lingui': linguiPlugin,
      '@nx': nxPlugin,
      'prefer-arrow': preferArrowPlugin,
      'import': importPlugin,
      'unused-imports': unusedImportsPlugin,
      'unicorn': unicornPlugin,
    },
    rules: {
      // General rules
      'func-style': ['error', 'declaration', { allowArrowFunctions: true }],
      'no-console': ['warn', { allow: ['group', 'groupCollapsed', 'groupEnd'] }],
      'no-control-regex': 0,
      'no-debugger': 'error',
      'no-duplicate-imports': 'error',
      'no-undef': 'off',
      'no-unused-vars': 'off',

      // Nx rules
      '@nx/enforce-module-boundaries': [
        'error',
        {
          enforceBuildableLibDependency: true,
          allow: [],
          depConstraints: [
            {
              sourceTag: 'scope:shared',
              onlyDependOnLibsWithTags: ['scope:shared'],
            },
            {
              sourceTag: 'scope:backend',
              onlyDependOnLibsWithTags: ['scope:shared', 'scope:backend'],
            },
            {
              sourceTag: 'scope:frontend',
              onlyDependOnLibsWithTags: ['scope:shared', 'scope:frontend'],
            },
            {
              sourceTag: 'scope:zapier',
              onlyDependOnLibsWithTags: ['scope:shared'],
            },
          ],
        },
      ],

      // Import rules
      'import/no-relative-packages': 'error',
      'import/no-useless-path-segments': 'error',
      'import/no-duplicates': ['error', { considerQueryString: true }],

      // Prefer arrow functions
      'prefer-arrow/prefer-arrow-functions': [
        'error',
        {
          disallowPrototype: true,
          singleReturnOnly: false,
          classPropertiesAllowed: false,
        },
      ],

      // Unused imports
      'unused-imports/no-unused-imports': 'warn',
      'unused-imports/no-unused-vars': [
        'warn',
        {
          vars: 'all',
          varsIgnorePattern: '^_',
          args: 'after-used',
          argsIgnorePattern: '^_',
        },
      ],
    },
  },

  // TypeScript specific configuration
  {
    files: ['**/*.ts'],
    languageOptions: {
      parser: typescriptParser,
      parserOptions: {
        project: [path.resolve(__dirname, 'tsconfig.*.json')],
      },
    },
    plugins: {
      '@typescript-eslint': typescriptEslint,
    },
    rules: {
      // TypeScript rules
      '@typescript-eslint/ban-ts-comment': 'error',
      '@typescript-eslint/consistent-type-imports': [
        'error',
        { 
          prefer: 'type-imports',
          fixStyle: 'inline-type-imports'
        },
      ],
      '@typescript-eslint/explicit-function-return-type': 'off',
      '@typescript-eslint/explicit-module-boundary-types': 'off',
      '@typescript-eslint/interface-name-prefix': 'off',
      '@typescript-eslint/no-empty-interface': [
        'error',
        {
          allowSingleExtends: true,
        },
      ],
      '@typescript-eslint/no-explicit-any': 'off',
      '@typescript-eslint/no-empty-function': 'off',
      '@typescript-eslint/no-unused-vars': 'off',

      // Nx dependency checks
      '@nx/dependency-checks': 'error',
    },
  },

  // JavaScript specific configuration
  {
    files: ['*.{js,jsx}'],
    rules: {
      // JavaScript-specific rules if needed
    },
  },

  // Test files
  {
    files: [
      '*.spec.@(ts|tsx|js|jsx)',
      '*.integration-spec.@(ts|tsx|js|jsx)',
      '*.test.@(ts|tsx|js|jsx)',
    ],
    languageOptions: {
      globals: {
        jest: true,
        describe: true,
        it: true,
        expect: true,
        beforeEach: true,
        afterEach: true,
        beforeAll: true,
        afterAll: true,
      },
    },
    rules: {
      '@typescript-eslint/no-non-null-assertion': 'off',
    },
  },

  // Constants files
  {
    files: ['**/constants/*.ts', '**/*.constants.ts'],
    rules: {
      '@typescript-eslint/naming-convention': [
        'error',
        {
          selector: 'variable',
          format: ['UPPER_CASE'],
        },
      ],
      'unicorn/filename-case': [
        'warn',
        {
          cases: {
            pascalCase: true,
          },
        },
      ],
      '@nx/workspace-max-consts-per-file': ['error', { max: 1 }],
    },
  },

  // JSON files
  {
    files: ['**/*.json'],
    languageOptions: {
      parser: jsoncParser,
    },
  },
];
