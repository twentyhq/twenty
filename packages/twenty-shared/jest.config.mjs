const jestConfig = {
  silent: true,
  displayName: 'twenty-shared',
  preset: '../../jest.preset.js',
  testEnvironment: 'jsdom',
  transformIgnorePatterns: ['../../node_modules/'],
  transform: {
    '^.+\\.[tj]sx?$': [
      '@swc/jest',
      {
        jsc: {
          parser: { syntax: 'typescript', tsx: true },
          transform: { react: { runtime: 'automatic' } },
        },
      },
    ],
  },
  moduleNameMapper: {
    // TODO prastoin investigate not working with pathsToModuleNameMapper
    /*
      {
        '^@/(.*)\\.js$': './src/$1',
        '^@/(.*)$': './src/$1',
        '^(\\.{1,2}/.*)\\.js$': '$1'
      } // use esm true
      { '^@/(.*)$': './src/$1' } // useEsm false
    */
    '/^@/(.*)$/': './src/$1',
    '\\.(jpg|jpeg|png|gif|webp|svg|svg\\?react)$':
      '<rootDir>/__mocks__/imageMockShared.js',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  extensionsToTreatAsEsm: ['.ts', '.tsx'],
  coverageDirectory: './coverage',
  coverageThreshold: {
    global: {
      statements: 80,
      lines: 90,
      functions: 75,
    },
  },
};

export default jestConfig;
