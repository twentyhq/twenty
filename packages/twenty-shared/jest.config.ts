import { JestConfigWithTsJest } from 'ts-jest';

const jestConfig: JestConfigWithTsJest = {
  silent: true,
  displayName: 'twenty-shared',
  preset: '../../jest.preset.js',
  testEnvironment: 'jsdom',
  transformIgnorePatterns: ['../../node_modules/'],
  transform: {
    '^.+\\.[tj]sx?$': [
      '@swc/jest',
      {
        jsc: {
          parser: { syntax: 'typescript', tsx: true },
          transform: { react: { runtime: 'automatic' } },
        },
      },
    ],
  },
  moduleNameMapper: {
    // TODO prastoin investigate not working with pathsToModuleNameMapper
    /*
      {
        '^@/(.*)\\.js$': './src/$1',
        '^@/(.*)$': './src/$1',
        '^(\\.{1,2}/.*)\\.js$': '$1'
      } // use esm true
      { '^@/(.*)$': './src/$1' } // useEsm false
    */
    '/^@/(.*)$/': './src/$1',
    '\\.(jpg|jpeg|png|gif|webp|svg|svg\\?react)$':
      '<rootDir>/__mocks__/imageMock.js',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  extensionsToTreatAsEsm: ['.ts', '.tsx'],
  coverageDirectory: './coverage',
  coverageThreshold: {
    global: {
      statements: 100,
      lines: 100,
      functions: 100,
    },
  },
};

export default jestConfig;
