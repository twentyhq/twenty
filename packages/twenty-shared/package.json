{
  "name": "twenty-shared",
  "main": "dist/twenty-shared.cjs.js",
  "module": "dist/twenty-shared.esm.js",
  "license": "AGPL-3.0",
  "scripts": {
    "build": "preconstruct build"
  },
  "engines": {
    "node": "^18.17.1",
    "npm": "please-use-yarn",
    "yarn": "^4.0.2"
  },
  "devDependencies": {
    "@babel/preset-env": "^7.26.9",
    "@preconstruct/cli": "^2.8.12",
    "@prettier/sync": "^0.5.2",
    "@types/babel__preset-env": "^7",
    "babel-plugin-module-resolver": "^5.0.2",
    "glob": "^11.0.1",
    "tsx": "^4.19.3"
  },
  "dependencies": {
    "@sniptt/guards": "^0.2.0",
    "zod": "3.23.8"
  },
  "preconstruct": {
    "tsconfig": "tsconfig.lib.json",
    "entrypoints": [
      "./index.ts",
      "./constants/index.ts",
      "./testing/index.ts",
      "./translations/index.ts",
      "./types/index.ts",
      "./utils/index.ts",
      "./workspace/index.ts"
    ]
  },
  "files": [
    "dist",
    "constants",
    "testing",
    "translations",
    "types",
    "utils",
    "workspace"
  ]
}
