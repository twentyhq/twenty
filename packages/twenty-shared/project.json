{
  "name": "twenty-shared",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/twenty-shared/src",
  "projectType": "library",
  "tags": [
    "scope:shared"
  ],
  "targets": {
    "build": {
      "dependsOn": [
        "generateBarrels",
        "^build"
      ],
      "outputs": [
        "{projectRoot}/dist",
        "{projectRoot}/constants/package.json",
        "{projectRoot}/constants/dist",
        "{projectRoot}/testing/package.json",
        "{projectRoot}/testing/dist",
        "{projectRoot}/translations/package.json",
        "{projectRoot}/translations/dist",
        "{projectRoot}/types/package.json",
        "{projectRoot}/types/dist",
        "{projectRoot}/utils/package.json",
        "{projectRoot}/utils/dist",
        "{projectRoot}/workspace/package.json",
        "{projectRoot}/workspace/dist"
      ]
    },
    "generateBarrels": {
      "executor": "nx:run-commands",
      "cache": true,
      "inputs": [
        "production",
        "{projectRoot}/scripts/generateBarrels.ts"
      ],
      "outputs": [
        "{projectRoot}/src/index.ts",
        "{projectRoot}/src/*/index.ts",
        "{projectRoot}/package.json"
      ],
      "options": {
        "command": "tsx {projectRoot}/scripts/generateBarrels.ts"
      }
    },
    "typecheck": {},
    "test": {},
    "lint": {
      "options": {
        "lintFilePatterns": [
          "{projectRoot}/src/**/*.{ts,tsx,json}",
          "{projectRoot}/package.json"
        ],
        "reportUnusedDisableDirectives": "error"
      },
      "configurations": {
        "fix": {}
      }
    },
    "fmt": {
      "options": {
        "files": "src"
      },
      "configurations": {
        "fix": {}
      }
    }
  }
}
