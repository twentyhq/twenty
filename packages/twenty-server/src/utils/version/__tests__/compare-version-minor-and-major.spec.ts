import { EachTestingContext } from 'twenty-shared/testing';

import {
  CompareVersionMajorAndMinorReturnType,
  compareVersionMajorAndMinor,
} from 'src/utils/version/compare-version-minor-and-major';

type IsSameVersionTestCase = EachTestingContext<{
  version1: string;
  version2: string;
  expected?: CompareVersionMajorAndMinorReturnType;
  expectToThrow?: boolean;
}>;
describe('It should compare two versions with', () => {
  const beneathVersionTestCases: IsSameVersionTestCase[] = [
    {
      context: {
        version1: '1.0.0',
        version2: '1.1.0',
        expected: 'lower',
      },
      title: 'different minor version',
    },
    {
      context: {
        version1: '2.3.0',
        version2: '2.4.0',
        expected: 'lower',
      },
      title: 'different minor version',
    },
    {
      context: {
        version1: '0.1.0',
        version2: '0.2.0',
        expected: 'lower',
      },
      title: 'different minor version with zero major',
    },
    {
      context: {
        version1: '2.3.5',
        version2: '2.4.1',
        expected: 'lower',
      },
      title: 'different minor and patch versions',
    },
    {
      context: {
        version1: '1.0.0-alpha',
        version2: '1.1.0-beta',
        expected: 'lower',
      },
      title: 'different minor version with different pre-release tags',
    },
    {
      context: {
        version1: 'v1.0.0',
        version2: 'v1.1.0',
        expected: 'lower',
      },
      title: 'different minor version with v prefix',
    },
    {
      context: {
        version1: '2.0.0',
        version2: '42.42.42',
        expected: 'lower',
      },
      title: 'above version2',
    },
    {
      context: {
        version1: '2.0.0',
        version2: 'v42.42.42',
        expected: 'lower',
      },
      title: 'above version2 with v-prefix',
    },
  ];

  const sameVersionTestCases: IsSameVersionTestCase[] = [
    {
      context: {
        version1: '1.1.0',
        version2: '1.1.0',
        expected: 'equal',
      },
      title: 'exact same version',
    },
    {
      context: {
        version1: '1.1.0',
        version2: '1.1.42',
        expected: 'equal',
      },
      title: 'exact same major and minor but different patch version',
    },
    {
      context: {
        version1: 'v1.1.0',
        version2: 'v1.1.0',
        expected: 'equal',
      },
      title: 'exact same version with v prefix',
    },
    {
      context: {
        version1: '1.1.0-alpha',
        version2: '1.1.0-alpha',
        expected: 'equal',
      },
      title: 'exact same version with same pre-release tag',
    },
    {
      context: {
        version1: '0.0.1',
        version2: '0.0.1',
        expected: 'equal',
      },
      title: 'exact same version with all zeros',
    },
    {
      context: {
        version1: 'v1.1.0',
        version2: '1.1.0',
        expected: 'equal',
      },
      title: 'same version with different v-prefix',
    },
  ];

  const aboveVersionTestCases: IsSameVersionTestCase[] = [
    {
      context: {
        version1: 'v42.1.0',
        version2: '2.0.0',
        expected: 'higher',
      },
      title: 'above version',
    },
    {
      context: {
        version1: '42.42.42',
        version2: '2.0.0',
        expected: 'higher',
      },
      title: 'above version with prefix',
    },
  ];
  const invalidTestCases: IsSameVersionTestCase[] = [
    {
      context: {
        version1: 'invalid',
        version2: '1.1.0',
        expectToThrow: true,
      },
      title: 'invalid version1',
    },
    {
      context: {
        version1: '1.0.0',
        version2: 'invalid',
        expectToThrow: true,
      },
      title: 'invalid version2',
    },
    {
      context: {
        version1: '1.0',
        version2: '1.1.0',
        expectToThrow: true,
      },
      title: 'incomplete version1',
    },
  ];

  test.each([
    ...sameVersionTestCases,
    ...invalidTestCases,
    ...beneathVersionTestCases,
    ...aboveVersionTestCases,
  ])(
    ' $title',
    ({ context: { version1, version2, expected, expectToThrow = false } }) => {
      if (expectToThrow) {
        expect(() =>
          compareVersionMajorAndMinor(version1, version2),
        ).toThrowErrorMatchingSnapshot();
      } else {
        expect(compareVersionMajorAndMinor(version1, version2)).toBe(expected);
      }
    },
  );
});
