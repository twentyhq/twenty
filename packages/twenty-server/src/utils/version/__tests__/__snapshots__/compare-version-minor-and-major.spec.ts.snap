// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should compare two versions with  incomplete version1 1`] = `"Received invalid version: 1.0 1.1.0"`;

exports[`It should compare two versions with  invalid version1 1`] = `"Received invalid version: invalid 1.1.0"`;

exports[`It should compare two versions with  invalid version2 1`] = `"Received invalid version: 1.0.0 invalid"`;
